u_fingerprinter.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2017 Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// Fingerprinter is a struct largely for holding options for the FingerprintClientHello func
+type Fingerprinter struct {
+	// KeepPSK will ensure that PSK extensions are passed along into the resulting ClientHelloSpec
+	KeepPSK bool
+	// AllowBluntMimicry will ensure that unknown extensions are passed along into the resulting ClientHelloSpec
+	// WARNING: there could be numerous subtle issues with ClientHelloSpecs that are generated with this flag
+	AllowBluntMimicry bool
+}
+
+// FingerprintClientHello returns a ClientHelloSpec which is based on the ClientHello that is passed in as the data argument
+//
+// If the ClientHello passed in has extensions that are not recognized or cannot be handled, it will return a non-nil error and a nil *ClientHelloSpec value
+//
+// The data should be the ClientHello record as produced by crypto/tls.clientHelloMsg.marshal()
+// ie. it should not contain the full tls record, just the handshake message as outlined in https://tools.ietf.org/html/rfc5246#section-7.4
+func (f *Fingerprinter) FingerprintClientHello(data []byte) (*ClientHelloSpec, error) {
```

Comment:
Have you parsed the minTLSVers, maxTLSVers, which are part of ClientHelloSpec?
It's possible that the code is doing the right thing, given the behavior of [SetTLSVers()](https://github.com/refraction-networking/utls/blob/186025ac7b77465439618d1aeb2a5e444714d1cc/u_conn.go#L512), just wanted to double check.

Reply:
I'm actually a bit confused as to the `TLSVersMax`/`TLSVersMin` functionality. In terms of mimicry, am I right to assume that, supposing there is no supported versions extension, we'd want to set `minTLSVers = maxTLSVers = packetVersionField`?

The only reason I hesitate w/ that approach is that for the parrots, you specify them explicitly. Does that mean that Chrome version 70 could possibly be sending v1.0 client hellos a la https://github.com/refraction-networking/utls/blob/4ee624457953e3730be8cb012f0cc0450867d1c9/u_parrots.go#L71 ?

Reply:
fwiw w/ this current implementation, if the supported versions extension is present, the `ClientHelloSpec` will use those versions (because of the `SetTLSVers` function), otherwise it will default to `min = TLS 1.0, max = TLS 1.2`

Reply:
Ah ok thanks for that bit of primer. I'll make this change.